pkg/kv/kvserver/kvstorage/snaprecv/sst_snapshot_storage.go

```diff
@@ -138,21 +141,34 @@ func (s *SSTSnapshotStorageScratch) NewFile(
 	return f, nil
 }
 
-// WriteSST writes SST data to a file. The method closes
-// the provided SST when it is finished using it. If the provided SST is empty,
-// then no file will be created and nothing will be written.
-func (s *SSTSnapshotStorageScratch) WriteSST(ctx context.Context, data []byte) error {
+// WriteSST creates an SST populated with the given write function, and writes
+// it to a file. Does nothing if no data is written.
+func (s *SSTSnapshotStorageScratch) WriteSST(
+	ctx context.Context, write func(storage.Writer) error,
+) error {
 	if s.closed {
 		return errors.AssertionFailedf("SSTSnapshotStorageScratch closed")
 	}
-	if len(data) == 0 {
+
+	// TODO(itsbilal): Write to SST directly rather than buffer in a MemObject.
```

Comment:
This is a questionable TODO. We could move `s.NewFile()` from below right here, and pass `f` into `MakeIngestionSSTWriter()`, but it would run every key write through this path (unless the SST writer has some buffer of its own):

https://github.com/cockroachdb/cockroach/blob/63545b4d7dfd9ac76f62b4b0ddd580b000980309/pkg/kv/kvserver/kvstorage/snaprecv/sst_snapshot_storage.go#L229-L231

Leaving this TODO for now.